File: tests_wraps/src/main/java/com/github/klee0kai/test/car/di/wrapped/custom/wrappers/CarProviderWrapper.java
package com.github.klee0kai.test.car.di.wrapped.custom.wrappers;

import com.github.klee0kai.stone.annotations.wrappers.WrappersCreator;
import com.github.klee0kai.stone.wrappers.creators.ProviderWrapper;

import javax.inject.Provider;

@WrappersCreator(wrappers = {CarLazy.class, CarProvide.class})
public class CarProviderWrapper implements ProviderWrapper {

    @Override
    public <Wr, T> Wr wrap(Class<Wr> wrapperCl, Provider<T> originalProvider) {
        if (wrapperCl.equals(CarLazy.class)) {
            return (Wr) new CarLazy<T>(originalProvider::get);
        }
        if (wrapperCl.equals(CarProvide.class)) {
            return (Wr) new CarProvide<T>(originalProvider::get);
        }
        return null;
    }

}

File: tests_wraps_kotlin/src/main/kotlin/com/github/klee0kai/test/car/di/wrapped/custom/wrappers/CarProviderWrapper.kt
package com.github.klee0kai.test.car.di.wrapped.custom.wrappers

import com.github.klee0kai.stone.annotations.wrappers.WrappersCreator
import com.github.klee0kai.stone.weakref.Provider
import com.github.klee0kai.stone.wrappers.creators.ProviderWrapper
import kotlin.reflect.KClass

@WrappersCreator(wrappers = [CarLazy::class, CarProvide::class])
class CarProviderWrapper : ProviderWrapper {

    override fun <Wr : Any, T : Any> wrap(
        wrapperCl: KClass<Wr>,
        originalProvider: Provider<T?>?,
    ): Wr? {
        return when (wrapperCl) {
            CarLazy::class -> {
                CarLazy { originalProvider?.get() } as Wr
            }

            CarProvide::class -> {
                CarProvide<T> { originalProvider?.get() } as Wr
            }

            else -> null
        }
    }

}
